libsodium-net/SodiumCore.cs

```
@@ -1,64 +1,84 @@
+		}
 
-    /// <summary>
-    /// Returns the version of libsodium in use.
-    /// </summary>
-    /// <returns>
-    /// The sodium version string.
-    /// </returns>
-    public static string SodiumVersionString()
-    {
-      var ptr = SodiumLibrary.sodium_version_string();
+		/// <summary>
+		/// Gets a random number.
+		/// </summary>
+		/// <param name="upperBound">Integer between 0 and 2147483647.</param>
+		/// <returns>An unpredictable value between 0 and upperBound (excluded).</returns>
+		public static int GetRandomNumber(int upperBound) {
+			var randomNumber = SodiumLibrary.randombytes_uniform(upperBound);
 
-      return Marshal.PtrToStringAnsi(ptr);
-    }
+			return randomNumber;
+		}
 
-    /// <summary>Initialize libsodium.</summary>
-    /// <remarks>This only needs to be done once, so this prevents repeated calls.</remarks>
-    public static void Init()
-    {
-      if (!_isInit)
-      {
-        SodiumLibrary.init();
-        _isInit = true;
-      }
-    }
-  }
+		/// <summary>
+		/// Returns the version of libsodium in use.
+		/// </summary>
+		/// <returns>
+		/// The sodium version string.
+		/// </returns>
+		public static string SodiumVersionString() {
+			var ptr = SodiumLibrary.sodium_version_string();
+
+			return Marshal.PtrToStringAnsi(ptr);
+		}
+
+		public static void SaveUnmanagedDlls() {
+			var assembly = Assembly.GetExecutingAssembly();
+			var dir = Path.GetDirectoryName(new Uri(assembly.Location).LocalPath);
+			var resources = assembly.GetManifestResourceNames()
+				.Where(name => name.EndsWith(".dll", StringComparison.OrdinalIgnoreCase));
+			foreach (var res in resources) {
+				try {
+					var dllFile = Path.Combine(dir, Regex.Replace(res, @".*\.(?=[^.]+\.dll)", "", RegexOptions.IgnoreCase));
+					using (var dest = new FileStream(dllFile, FileMode.Create, FileAccess.Write))
+					using (var src = assembly.GetManifestResourceStream(res)) {
+						src.CopyTo(dest);
+						dest.Close();
+					}
+				} catch { }
```

This will fail silently in two cases that I can see:

1. If the user running this doesn't have write permission to the directory. This could happen in a number of situations, but the one that comes to mind first is ASP.NET applications - the permissions they run under is often limited, and this could prevent the application from working at all. We may need to give this scenario more thought, to ensure that we don't break existing applications. Thoughts?
2. The file is locked, which is a common issue when dealing with these unmanaged DLLs. This may or may not be an issue, depending on version. Perhaps it would make sense, on fail, to see if the hash of the file on disk matches what was to be written - if there's a match, then it doesn't matter, and if it doesn't, throw an exception. Thoughts?